src/ngraph/runtime/gpu/cudnn_emitter.cpp

```diff
@@ -0,0 +1,177 @@
+/*******************************************************************************
+* Copyright 2018 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+#include <iostream>
+#include <vector>
+
+#include "ngraph/runtime/gpu/cudnn_emitter.hpp"
+#include "ngraph/runtime/gpu/gpu_runtime_context.hpp"
+
+using namespace ngraph;
+using namespace ngraph::runtime::gpu;
+
+void CUDNNEmitter::invoke(size_t primitive_index,
+                          const std::vector<void*>& args,
+                          const std::vector<void*>& result)
+{
+    m_cudnn_primitives[primitive_index](args, result);
+}
+
+size_t CUDNNEmitter::register_primitive(
+    const std::function<void(std::vector<void*>, std::vector<void*>)>& f)
+{
+    // try emplace
+    m_cudnn_primitives.push_back(f);
```

Comment:
Is It possible two reduce share the same function?

Reply:
I think it's fine for now. but we need to revisit it in the future for optimization. We can do something similar to pool. If we can find in pool, we don't need to add another one. save some space and overhead.